rosbag2_py/src/rosbag2_py/_rosbag2_py.cpp

```diff
@@ -0,0 +1,256 @@
+#include "rosbag2_cpp/readers/sequential_reader.hpp"
+#include "rosbag2_cpp/reader.hpp"
+#include "rosbag2_cpp/storage_options.hpp"
+#include "rosbag2_cpp/writer.hpp"
+#include "rosbag2_cpp/writers/sequential_writer.hpp"
+#include "rosbag2_storage/ros_helper.hpp"
+#include "rosbag2_storage/storage_filter.hpp"
+#include "rosbag2_storage/topic_metadata.hpp"
+
+namespace rosbag2_py
+{
+
+class Reader
+{
+public:
+  Reader()
+  {
+    initialized_ = false;
+  }
+
+  void open(
+    rosbag2_cpp::StorageOptions & storage_options,
+    rosbag2_cpp::ConverterOptions & converter_options)
+  {
+    reader_->open(storage_options, converter_options);
+  }
+
+  bool has_next()
+  {
+    return reader_->has_next();
+  }
+
+  /// Return a tuple containing the topic name, the serialized ROS message, and
+  /// the timestamp.
+  pybind11::tuple read_next()
+  {
+    const auto next = reader_->read_next();
+    rcutils_uint8_array_t rcutils_data = *next->serialized_data.get();
+    std::string serialized_data(rcutils_data.buffer,
+      rcutils_data.buffer + rcutils_data.buffer_length);
+    return pybind11::make_tuple(
+      next->topic_name, pybind11::bytes(serialized_data), next->time_stamp);
+  }
+
+  /// Return a mapping from topic name to topic type.
+  std::vector<rosbag2_storage::TopicMetadata> get_all_topics_and_types()
+  {
+    return reader_->get_all_topics_and_types();
+  }
+
+  void set_filter(const rosbag2_storage::StorageFilter & storage_filter)
+  {
+    return reader_->set_filter(storage_filter);
+  }
+
+  void reset_filter()
+  {
+    reader_->reset_filter();
+  }
+
+protected:
+  std::unique_ptr<rosbag2_cpp::Reader> reader_;
+  bool initialized_;
+};
+
+class SequentialReader : public Reader
+{
+public:
+  SequentialReader()
+  : Reader()
+  {
+    auto sequential_reader = std::make_unique<rosbag2_cpp::readers::SequentialReader>();
+    reader_ = std::make_unique<rosbag2_cpp::Reader>(std::move(sequential_reader));
+    initialized_ = true;
+  }
+};
+
+class SequentialCompressionReader : public Reader
+{
+public:
+  SequentialCompressionReader()
+  : Reader()
+  {
+    auto sc_reader = std::make_unique<rosbag2_compression::SequentialCompressionReader>();
+    reader_ = std::make_unique<rosbag2_cpp::Reader>(std::move(sc_reader));
+    initialized_ = true;
+  }
+};
+
+class Writer
+{
+public:
+  Writer()
+  {
+    initialized_ = false;
+  }
+
+  void open(
+    rosbag2_cpp::StorageOptions & storage_options,
+    rosbag2_cpp::ConverterOptions & converter_options)
+  {
+    writer_->open(storage_options, converter_options);
+  }
+
+  void create_topic(const rosbag2_storage::TopicMetadata & topic_with_type)
+  {
+    writer_->create_topic(topic_with_type);
+  }
+
+  void remove_topic(const rosbag2_storage::TopicMetadata & topic_with_type)
+  {
+    writer_->remove_topic(topic_with_type);
+  }
+
+  /// Write a serialized message to a bag file
+  void write(
+    const std::string & topic_name, const std::string & message,
```

Comment:
why is the message encapsulated as a `std::string` here?

Reply:
The Python object being passed in is `bytes` type, and `std::string` or `char *` will be passed in without conversion
https://pybind11.readthedocs.io/en/stable/advanced/cast/strings.html#passing-bytes-to-c Here since it needs to be passed to `rosbag2_storage::make_serialized_message`, which takes the size, `std::string` is easier to get the size. `pybind11:;bytes` is another option, but `rosbag2_storage::make_serialized_message` also takes a `const void *`, and using the `char *` from a string is fewer casting / conversions.